### PR TITLE
fix: do not build OIDC config unless enabled

### PR DESCRIPTION
**Pull request by Cali0707**
Temporary fix for #3901 (this does not fix the underlying problem as far as I can tell, but will enable anyone not interested in OIDC to use 1.14+)

## Proposed Changes

- Do not load the OIDC config unless it is enabled
- Refactor the receiver `Main` class to allow for redeploying the receiver verticles when the OIDC config changes
- Set an annotation on the receiver pods to force k8s to update the files mounted from the `config-features` configmap, so that the receiver quickly picks up on changes to feature flags.

**Release Note**

```release-note
The receiver components now successfully start on EKS. Special shoutout to @treyhyde for help debugging this.
```

### CODE REVIEW

**Comment by Cali0707** [comment]
TODO (as a follow up): refactor the `FileWatcher` class to be able to either watch a single file or a directory. Because of how configmaps are mounted as files into pods this PR just watches the specific file associated with the OIDC feature flag. But, going forwards we will probably end up wanting to support more feature flags, so we should make the `FileWatcher` class more generic there

**Comment by Cali0707** [comment]
/cc @pierDipi 

This should be good to go now, I updated it to not redeploy the verticles (like we had discussed offline).

Now, we are using a listener + callback pattern for the OIDC Discovery stuff

**Comment by Cali0707** [comment]
/cc @pierDipi @creydr @Leo6Leo 

I have fixed the merge conflicts here

**Comment by Cali0707** [comment]
/hold

Looks like the tests are failing for some reason...

**Comment by creydr** [comment]
`/home/prow/go/src/knative.dev/eventing-kafka-broker is out of date. Please run hack/update-codegen.sh`

@Cali0707 can you run it?

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/oidc/OIDCDiscoveryConfigListener.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        return oidcDiscoveryConfig;
+    }
+
+    public int registerCallback(Consumer<OIDCDiscoveryConfig> callback) {
+        if (this.callbacks == null) {
+            this.callbacks = new ArrayList<>();
+        }
+
+        this.callbacks.add(callback);
+        return this.callbacks.size();
```
Comment (creydr) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/oidc/OIDCDiscoveryConfigListener.java:74]: IIUC you are using this return value as some "id", which you later pass to "deregisterCallback()". This uses `.set(id, null)` to remove it.
So you do the following on an initial empty arraylist
1. add a callback
2. return as "callbackId" `1`
3. later want to deregister the callbackId `1`, but actually you would need to call `deregisterCallback(0)`

sry, but I didn't find a way to explain it more complicated :facepalm: 

tl;dr; is, that IIUC, you should return here `this.callbacks.size() - 1`, if you want to use this as the "callbackId"

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java [hunk 2]
```diff
@@ -135,8 +136,8 @@ public void start(final Promise<Void> startPromise) {
                 .watchIngress(IngressReconcilerListener.all(this.ingressProducerStore, this.ingressRequestHandler))
                 .buildAndListen(vertx);
 
-        TokenVerifier tokenVerifier = new TokenVerifierImpl(vertx, oidcDiscoveryConfig);
-        this.authenticationHandler = new AuthenticationHandler(tokenVerifier);
+        this.buildAuthHandler(oidcDiscoveryConfigListener.getOidcDiscoveryConfig());
+        this.oidcDiscoveryCallbackId = this.oidcDiscoveryConfigListener.registerCallback(this::buildAuthHandler);
```
Comment (creydr) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:140]: Maybe we can add a quick note somewhere, that the FileWatcher runs the callbacks, when it gets started (in it's .run() ), that way we can be "sure", that `this.authenticationHandler` is not null (which is used in line 243)

**Review by creydr — COMMENTED** [review]
Thanks for working on this :muscle: . Did a first round. But I probably need to run it locally too :D

**Comment by Cali0707** [comment]
/cc @creydr 

Thanks for taking a first look!

**Comment by Cali0707** [comment]
/unhold

Seems like things are passing now, thanks for finding that off by one error @creydr !!

**Comment by Cali0707** [comment]
/test upgrade-tests

Last failure looks like a flake, cert-manager failed to come ready
